### PR TITLE
Add Environment

### PR DESCRIPTION
**Pull request by cazfletch**
Add new environment to list

closes to #1086

Signed-off-by: Caroline Fletcher <caroline.fletcher@uk.ibm.com>

### CODE REVIEW

File: client/cucumber/features/fabric-environments.feature [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        Then there should be a tree item with a label '<label>' in the 'Fabric Environments' panel
+        Then the tree item should have a tooltip equal to '<tooltip>'
+        When I start the Local Fabric
+        And the 'Local Fabric' environment is connected
+        Then there should be a instantiated smart contract tree item with a label '<instantiatedName>' in the 'Fabric Environments' panel
+        Examples:
+        | language   | assetType | name               | instantiatedName         | version | label                            | tooltip                                                                    |
+        | JavaScript | Conga     | JavaScriptContract | JavaScriptContract@0.0.2 | 0.0.2   | Local Fabric  ○ (click to start) | Creates a local development runtime using Hyperledger Fabric Docker images |
+
+    Scenario Outline: After tear down and start there are no smart contracts
```
Comment (Jakeeyturner) [inline client/cucumber/features/fabric-environments.feature]: Please could you change this to `teardown` to make it consistent with everything else?

File: client/cucumber/features/fabric-environments.feature [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        Given the Local Fabric is running
+        And the 'Local Fabric' environment is connected
+        And a <language> smart contract for <assetType> assets with the name <name> and version <version> 
+        And the contract has been created
+        And the contract has been packaged
+        And the package has been installed
+        And the contract has been instantiated with the transaction '' and args '', not using private data
+        When I tear down the Local Fabric
```
Comment (Jakeeyturner) [inline client/cucumber/features/fabric-environments.feature]: Please could you change this to `teardown` to make it consistent with everything else?

File: client/cucumber/features/fabric-environments.feature [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    Scenario: It should create an envrionment
+        When I create an environment 'myFabric'
+        Then there should be a tree item with a label 'myFabric' in the 'Fabric Environments' panel
+        And the tree item should have a tooltip equal to 'myFabric'
+
+    @otherFabric
+    Scenario Outline: It should connect to an environment
+        Given an environment 'myFabric' exists
+        And the wallet 'myWallet' with identity 'conga' and mspid 'Org1MSP' exists
+        When I connect to the environment 'myFabric'
```
Comment (Jakeeyturner) [inline client/cucumber/features/fabric-environments.feature]: Should be be explicit with this step and make it something like:
`When I connect to the environment 'myFabric' with the wallet 'myWallet' and identity 'conga'`.

If we do something like that we can eventually add tests such as:

```
...
And the wallet 'myWallet' with identities 'conga1', 'conga2', 'conga3' and mspid 'Org1MSP' exists
When I connect to the environment 'myFabric' with the wallet 'myWallet' and identity 'conga2'
...
Then I should be connected to 'myFabric' with the identity 'conga2'
```

File: client/cucumber/features/upgrade.feature [hunk 4]
```diff
@@ -13,7 +13,7 @@ Feature: Upgrade Smart Contracts
         And the contract has been packaged
         And the package has been installed
         When I upgrade the installed package with the transaction '' and args '', not using private data
-        Then there should be a instantiated smart contract tree item with a label '<upgradedName>' in the 'Local Fabric Ops' panel
+        Then there should be a instantiated smart contract tree item with a label '<upgradedName>' in the 'Fabric Environments' panel
```
Comment (Jakeeyturner) [inline client/cucumber/features/upgrade.feature]: For grammatical purposes, we should make this:
_"Then there should be a**n** instantiated..."_

File: client/src/commands/UserInputUtil.ts [hunk 5]
```diff
@@ -73,6 +73,9 @@ export class UserInputUtil {
     static readonly ADD_LOCAL_ID_SECRET_OPTION: string = 'Provide an enrollment ID and secret';
     static readonly ADD_JSON_ID_OPTION: string = 'Provide a JSON identity file';
 
+    static readonly ADD_MORE_NODES: string = 'Add more (json) node definitions';
```
Comment (Jakeeyturner) [inline client/src/commands/UserInputUtil.ts]: Can we change json to JSON please

File: client/src/commands/addEnvironmentCommand.ts [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        const fabricEnvironmentRegistry: FabricEnvironmentRegistry = FabricEnvironmentRegistry.instance();
+
+        const environmentName: string = await UserInputUtil.showInputBox('Enter a name for the environment');
+        if (!environmentName) {
+            return;
+        }
+
+        if (fabricEnvironmentRegistry.exists(environmentName) || environmentName === FabricRuntimeUtil.LOCAL_FABRIC) {
+            // Environment already exists
+            throw new Error('A environment with this name already exists.');
```
Comment (Jakeeyturner) [inline client/src/commands/addEnvironmentCommand.ts]: Should be 'An environment'

File: client/src/commands/addEnvironmentCommand.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        const homeExtDir: string = UserInputUtil.getDirPath(dirPath);
+        const environmentPath: string = path.join(homeExtDir, 'environments', environmentName, 'nodes');
+
+        await fs.ensureDir(environmentPath);
+        for (const nodeUri of nodeUris) {
+            try {
+                const destPath: string = path.join(environmentPath, path.basename(nodeUri.fsPath));
+                await fs.copyFile(nodeUri.fsPath, destPath);
+            } catch (error) {
+                outputAdapter.log(LogType.ERROR, `error importing node file ${nodeUri.fsPath}: ${error.message}`, `error importing node file ${nodeUri.fsPath}: ${error.toString()}`);
```
Comment (Jakeeyturner) [inline client/src/commands/addEnvironmentCommand.ts]: Can we make 'error' 'Error' please (in both args)

File: client/src/commands/createNewIdentityCommand.ts [hunk 8]
```diff
@@ -60,7 +60,7 @@ export async function createNewIdentity(certificateAuthorityTreeItem?: Certifica
         const wallet: IFabricWallet = await connection.getWallet(certificateAuthorityName);
         const identityExists: boolean = await wallet.exists(identityName);
         if (identityExists) {
-            outputAdapter.log(LogType.ERROR, `An identity called ${identityName} already exists in the runtime wallet`, `An identity called ${identityName} already exists in the runtime wallet`);
+            outputAdapter.log(LogType.ERROR, `An identity called ${identityName} already exists`, `An identity called ${identityName} already exists`);
```
Comment (Jakeeyturner) [inline client/src/commands/createNewIdentityCommand.ts]: Can you remove the last arg (shouldn't need it as the message it the same)

File: client/cucumber/features/fabric-environments.feature (hunk 3 above)
Comment (cazfletch) [reply client/cucumber/features/fabric-environments.feature]: don't need this

File: client/src/commands/installCommand.ts [hunk 9]
```diff
@@ -76,7 +76,10 @@ export async function installSmartContract(treeItem?: BlockchainTreeItem, peerNa
             title: 'IBM Blockchain Platform Extension',
             cancellable: false
         }, async (progress: vscode.Progress<{ message: string }>) => {
-            VSCodeBlockchainDockerOutputAdapter.instance().show();
+
+            if (FabricEnvironmentManager.instance().getEnvironmentRegistryEntry().managedRuntime) {
```
Comment (Jakeeyturner) [inline client/src/commands/installCommand.ts]: Could you split this out to something like?
```
const environmentRegistryEntry: FabricEnvironmentRegistryEntry = FabricEnvironmentManager.instance().getEnvironmentRegistryEntry();
if(environmentRegistryEntry.managedRuntime){
    ...
}
```

```

File: client/src/commands/instantiateCommand.ts [hunk 10]
```diff
@@ -172,15 +172,17 @@ export async function instantiateSmartContract(treeItem?: BlockchainTreeItem, ch
 
             progress.report({ message: 'Instantiating Smart Contract' });
 
-            VSCodeBlockchainDockerOutputAdapter.instance().show();
+            if (FabricEnvironmentManager.instance().getEnvironmentRegistryEntry().managedRuntime) {
```
Comment (Jakeeyturner) [inline client/src/commands/instantiateCommand.ts]: Can you split it here as well please

File: client/src/commands/submitTransaction.ts [hunk 11]
```diff
@@ -127,7 +127,11 @@ export async function submitTransaction(evaluate: boolean, treeItem?: Instantiat
         try {
             progress.report({ message: `${actioning} transaction ${transactionName}` });
             outputAdapter.log(LogType.INFO, undefined, `${actioning} transaction ${transactionName} with args ${args}`);
-            VSCodeBlockchainDockerOutputAdapter.instance().show();
+
+            if (FabricConnectionManager.instance().getGatewayRegistryEntry().managedRuntime) {
```
Comment (Jakeeyturner) [inline client/src/commands/submitTransaction.ts]: Could you split this out please

File: client/src/commands/upgradeCommand.ts [hunk 12]
```diff
@@ -195,13 +196,17 @@ export async function upgradeSmartContract(treeItem?: BlockchainTreeItem, channe
 
             progress.report({ message: 'Upgrading Smart Contract' });
 
+            if (FabricEnvironmentManager.instance().getEnvironmentRegistryEntry().managedRuntime) {
```
Comment (Jakeeyturner) [inline client/src/commands/upgradeCommand.ts]: Can you split this please

**Comment by Jakeeyturner** [comment]
Would it also be possible to change the message when connecting to the environment? (At the moment, connecting to the Local Fabric environment logs/shows `"Connected to local_fabric"`)

**Pull request merged**